When a signed request fails, fire an action so that sites can log the details of the failure

mdawaffe opened the pull request:
#### Changes proposed in this Pull Request:
Adds a `jetpack_verify_signature_error` action that is fired when an XML-RPC or REST API request's signature is invalid.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
New feature: ability to do better error reporting.

#### Testing instructions:
WIP

#### Proposed changelog entry for your changes:
None required?

mdawaffe committed:
Jetpack Signature Verification: Send Errors back in HTTP Response

WIP

mdawaffe committed:
Send new error

eliorivero commented on class.jetpack.php:
```diff
@@ -5395,6 +5422,18 @@ function wp_rest_authenticate( $user ) {
+		header( sprintf(
+			'X-Jetpack-Signature-Error: %s',
+			base64_encode( json_encode( $error_data['signature_details'] ) )
+		) );
```
Can you simplify the logic here so the function doesn't return void on one case and nothing on the other?
```
		if ( isset( $error_data['signature_details'] ) ) {
			header( sprintf(
				'X-Jetpack-Signature-Error: %s',
				base64_encode( json_encode( $error_data['signature_details'] ) )
			) );
		}
```

gravityrail replied:
Why does it matter what this function returns? We are not checking the output anywhere.

Nested if's don't really simplify the logic IMO, they just move it around.

mdawaffe replied:
There is no practical difference between `return;` and not calling `return`. Both return `null`:

```php
function return_nothing() {
  return;
}

function do_not_return() {
   // nothing
}

function return_null() {
  return null;
}

var_dump( return_nothing(), do_not_return(), return_null() ); // NULL, NULL, NULL
```

The only difference between `void` and `null` occurs when using `void` as a return type hint.

```php
function this_works() : void {
  return; // fine
}

function this_also_works() : void {
   // nothing - also fine
}

function this_breaks() : void {
  return null; // not fine
}
```

The last function will cause a parse error.

So PHP knows the difference, but calling code can't. (Well - there might be a way to tell the difference with reflection, but that's silly :))

In the case of this `send_signature_error_header()` method, if we were to include a `void` return type hint, everything would be fine, so my claim is that this falls into the "there is no practical difference" category :)

I think an aesthetic/code style argument could be made, but I (subjectively) disagree ;) I prefer the "bail as early as possible and indent as little as possible" argument in this case.

mdawaffe replied:
(Moot now because the code is gone ;))

gravityrail commented:
I dig the approach, haven't tested interactively myself but seems like a good idea.

I was initially skeptical about always returning the header if the signature is invalid, since some endpoints don't require a valid signature and therefore it doesn't matter, but I'm coming around to this way of doing things.

mdawaffe committed:
Don't send signature error details.

Just call an action we can use for logging.

mdawaffe commented:
I've thought about this some more, and I don't like the idea of exposing the exact error state; I'm worried it leaks too much information. We may want to do this eventually, but I suggest being more conservative for now: Instead of always responding with exactly what failed, just fire an action.

We can add logging or use the header on Atomic sites to gather some data in a more controlled fashion than we could if we turned this on everywhere.

mdawaffe committed:
typo

mdawaffe committed:
Another typo.

mdawaffe committed:
So many typos :)

mdawaffe committed:
Check `isset()` before using.

mdawaffe committed:
Fix tests.

gravityrail reviewed (dismissed):
This looks good to me. Merge when ready.

zinigor reviewed (dismissed):
LGTM, removing mixed signals, like that giant DO NOT MERGE label.

mdawaffe committed:
Add provided and expected signatures to signature details.

mdawaffe committed:
Inline comment about safety of exposing expected signature.

mdawaffe commented:
I just added some more debugging details: the provided signature and the expected signature (when we know it).

Prior to this latest commit, I was reasonably sure all this debugging information could be made public. I switched from the original method of always sending the debugging information back in the response to the new method of just firing an action so that the debugging can be logged someplace custom out of caution: I believe the debugging data is safe, but if it does expose a vulnerability, better that we can control it without having to ship a new version of the plugin.

After this latest commit, I no longer believe the debugging information is safe to expose publicly. Giving direct feedback about what signature was expected in general seems very sketchy to me. For this specific case, it would be disastrous without a rewrite of how we handle things like duplicate requests.

If we ever do want to make this data publicly available, we'll need to be careful.

mdawaffe commented:
I'm still thinking a couple things through :) Will be ready for review shortly.

mdawaffe committed:
Use up the nonce before we report on signature match.

This makes the signature debugging details less sensitive (but probably
not non-sensitive).

mdawaffe commented:
906206aae89f00efd2922f6d4f3eb57771cee068 makes the debugging info safer (not safe :)) to expose publicly. The request duplication detection code (the nonce stuff) is now run before we add the expected signature to the debugging details.

jeherve commented on class.jetpack.php:
```diff
@@ -5165,23 +5164,55 @@ public function reset_saved_auth_state() {
+
+			if ( is_wp_error( $this->xmlrpc_verification ) ) {
+				// Action for logging signature verification errors. This data is sensitive.
+				do_action( 'jetpack_verify_signature_error', $this->xmlrpc_verification );
```
Could you add a docblock for that action, to make our parser happy?

gravityrail committed:
Add docblock for signature verification error action

gravityrail commented:
I added that missing docblock

gravityrail committed:
Additional error statuses based on parsing token

zinigor committed:
Merge branch 'master' into add/signature-error-reporting

zinigor reviewed (dismissed):
Looks like this one is going to be green: I have merged the conflicting changes back to this branch. Feel free to merge this one!

gravityrail committed:
Add a small amount of extra information in the error

gravityrail committed:
Merge

gravityrail committed:
Slightly better name for error

jeherve reviewed (approved):
This works for me. Merging.
